SignalMessaging/Models/OWSQuotedReplyModel.m

```diff
@@ -0,0 +1,183 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSQuotedReplyModel.h"
+#import <SignalServiceKit/MIMETypeUtil.h>
+#import <SignalServiceKit/OWSMessageSender.h>
+#import <SignalServiceKit/TSAccountManager.h>
+#import <SignalServiceKit/TSAttachmentStream.h>
+#import <SignalServiceKit/TSIncomingMessage.h>
+#import <SignalServiceKit/TSMessage.h>
+#import <SignalServiceKit/TSOutgoingMessage.h>
+#import <SignalServiceKit/TSQuotedMessage.h>
+#import <SignalServiceKit/TSThread.h>
+
+// View Model which has already fetched any thumbnail attachment.
+@implementation OWSQuotedReplyModel
+
+- (instancetype)initWithTimestamp:(uint64_t)timestamp
+                         authorId:(NSString *)authorId
+                             body:(NSString *_Nullable)body
+                 attachmentStream:(nullable TSAttachmentStream *)attachmentStream
+{
+    return [self initWithTimestamp:timestamp
+                          authorId:authorId
+                              body:body
+                    thumbnailImage:attachmentStream.thumbnailImage
+                       contentType:attachmentStream.contentType
+                    sourceFilename:attachmentStream.sourceFilename
+                  attachmentStream:attachmentStream];
+}
+
+
+- (instancetype)initWithTimestamp:(uint64_t)timestamp
+                         authorId:(NSString *)authorId
+                             body:(nullable NSString *)body
+                   thumbnailImage:(nullable UIImage *)thumbnailImage
+                      contentType:(nullable NSString *)contentType
+                   sourceFilename:(nullable NSString *)sourceFilename
+                 attachmentStream:(nullable TSAttachmentStream *)attachmentStream
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _timestamp = timestamp;
+    _authorId = authorId;
+    _body = body;
+    _thumbnailImage = thumbnailImage;
+    _contentType = contentType;
+    _sourceFilename = sourceFilename;
+    _attachmentStream = attachmentStream;
+
+    return self;
+}
+
+- (instancetype)initWithQuotedMessage:(TSQuotedMessage *)quotedMessage
+                          transaction:(YapDatabaseReadTransaction *)transaction
+{
+    OWSAssert(quotedMessage.quotedAttachments.count <= 1);
+    OWSAttachmentInfo *attachmentInfo = quotedMessage.quotedAttachments.firstObject;
+
+    UIImage *_Nullable thumbnailImage;
+    if (attachmentInfo.thumbnailAttachmentStreamId) {
+        TSAttachment *attachment =
+            [TSAttachment fetchObjectWithUniqueID:attachmentInfo.thumbnailAttachmentStreamId transaction:transaction];
+
+        TSAttachmentStream *attachmentStream;
+        if ([attachment isKindOfClass:[TSAttachmentStream class]]) {
+            attachmentStream = (TSAttachmentStream *)attachment;
+            thumbnailImage = attachmentStream.image;
```

Comment:
This looks like it's eagerly loading into memory all thumbnails for all quoted messages as soon as the `ConversationViewItem` is created.  I suspect we'll want to do that lazily.  In fact, to support browsing _far_ back into conversation histories, I suspect we'll want to eventually unload these thumbnails if they are _far_ offscreen.

Reply:
That's a good plan. 

Since this behavior is essentially the same as before when we were passing around the raw `(NSData *)thumbnailData`, and likely entails changes in the view layer to request this data be loaded/unloaded I'd like to leave this proposed enhancement for a separate PR.